src/main/java/org/datadog/jenkins/plugins/datadog/clients/DatadogHttpClient.java

```diff
@@ -76,28 +90,51 @@ public boolean sendEvent(JSONObject payload) {
 
     @Override
     public void incrementCounter(String name, String hostname, JSONArray tags) {
-        ConcurrentMap<CounterMetric, Integer> counters = ConcurrentMetricCounters.Counters.get();
-        CounterMetric counterMetric = new CounterMetric(tags, name, hostname);
-        Integer previousValue = counters.putIfAbsent(counterMetric, 1);
-        if (previousValue != null){
-            boolean ok = counters.replace(counterMetric, previousValue, previousValue + 1);
-            while(!ok) {
-                previousValue = counters.get(counterMetric);
-                ok = counters.replace(counterMetric, previousValue, previousValue + 1);
+        ConcurrentMetricCounters.lock.lock();
+        try {
+            ConcurrentMetricCounters countersInstance = ConcurrentMetricCounters.getInstance();
+            ConcurrentMap<CounterMetric, Integer> counters = countersInstance.get();
+            CounterMetric counterMetric = new CounterMetric(tags, name, hostname);
+            Integer previousValue = counters.putIfAbsent(counterMetric, 1);
+            if (previousValue != null){
+                boolean ok = counters.replace(counterMetric, previousValue, previousValue + 1);
+                // NOTE:
+                // This while loop below should never be called since we are using a lock when flushing and
+                // incrementing counters.
+                while(!ok) {
+                    logger.warning("Couldn't increment counter " + name + " with value " + (previousValue + 1) +
+                            " previousValue = " + previousValue);
+                    previousValue = counters.get(counterMetric);
+                    ok = counters.replace(counterMetric, previousValue, previousValue + 1);
+                }
             }
+            previousValue = previousValue == null ? 0 : previousValue;
+            logger.fine("Counter " + name + " updated from previousValue " + previousValue + " to "
+                    + (previousValue + 1));
```

Comment:
Why not make this a synchronized method of the `ConcurrentMetricCounters` object like , and remove that manual lock ?

Reply:
synchronized would prevent the same method to be called at the same time. Here with my lock I prevent we don't call both method at the same time

Reply:
Yeah that's my point, if you had a synchronized method that did the entire thing, you'd be fine, same idea as the `getAndReset`. 

So IMO, you shouldn't even be able to manually update the counters instance from outside of the class. 
The only way for you to obtain and manipulate the counters, should be when you are sure you will not do anymore operations on them, and so you would call `getAndReset` and do whatever you want with it at this point.

IOW, no `get` method, only a synchronized `increment` and a synchronized `getAndReset`. WDYT ? I believe it would make everything safer.